aten/src/ATen/native/TensorIterator.cpp

```diff
@@ -894,20 +894,67 @@ int TensorIterator::get_dim_to_split() const {
   return dim_to_split;
 }
 
-void TensorIterator::fast_set_up() {
-  //this function is called if all the inputs are contiguous to avoid needless reordering of dimensions
-  //and tracking output strides
-  //TODO enable fast handling for reductions
-  //TODO enable fast handling for channels_last
-
-  //allocate contiguous tensor for output
-  for (int i = 0; i < num_outputs_; i++){
-      auto& op = operands_[i];
-      if (!op.tensor.defined()) {
-        TORCH_INTERNAL_ASSERT(op.is_type_defined(), "no type for operand", i);
-        op.tensor = at::empty(shape_, op.options());
+bool TensorIterator::fast_set_up() {
+  // This function tries to do a fast setup to avoid needless reordering of dimensions and tracking output strides
+  // Return true if it can do fast setup or false otherwise
+  // TODO enable fast handling for reductions
+  FastSetupType setup_type = compute_fast_setup_type();
+  if (setup_type == FastSetupType::NONE) {
+    return false;
+  }
+
+  // allocate memory for output, memory format depends on setup_type
+  switch (setup_type) {
+    case FastSetupType::CONTIGUOUS:
+      {
+        for (int i = 0; i < num_outputs_; i++){
+          auto& op = operands_[i];
+          if (!op.tensor.defined()) {
+            TORCH_INTERNAL_ASSERT(op.is_type_defined(), "no type for operand", i);
+            op.tensor = at::empty(shape_, op.options(), MemoryFormat::Contiguous);
+            op.current_dtype = op.target_dtype;
+          }
+        }
+        break;
       }
-      op.current_dtype = op.target_dtype;
+    case FastSetupType::CHANNELS_LAST:
+      {
+        for (int i = 0; i < num_outputs_; i++){
+          auto& op = operands_[i];
+          if (!op.tensor.defined()) {
+            TORCH_INTERNAL_ASSERT(op.is_type_defined(), "no type for operand", i);
+            op.tensor = at::empty(shape_, op.options(), MemoryFormat::ChannelsLast);
+            op.current_dtype = op.target_dtype;
+          }
+        }
+        break;
+      }
+    case FastSetupType::NON_OVERLAPPING_DENSE:
+      {
+        // find the index of a defined tensor in operands_ start from input tensor
+        int i_defined;
+        for (i_defined = ntensors() - 1; i_defined >= 0; --i_defined) {
+          if (operands_[i_defined].tensor.defined()) break;
+        }
+        TORCH_CHECK(i_defined >= 0, "Can not find a defined tensor when fast allocating memory to outputs");
+        for (int i = 0; i < num_outputs_; i++){
```

Comment:
ntensors()

Reply:
My bad to give you wrong info before. We should check output only, but in else if statement, I shouldn't check op.is_output again.